### DISCUSSION TITLE
generic/dynamic model fourier series

### DISCUSSION THREAD

**Discussion by bytelinker** [lmfit/lmfit-py#914] [General]
Hello,

which is the best way to set up a generic or dynamic fourier series fit model?
The degree of fourier coeffcients should be variable or could be determined by the numbers of guess parameters.

Starting points are:
https://stackoverflow.com/questions/52524919/fourier-series-fit-in-python
https://symfit.readthedocs.io/en/stable/examples/ex_fourier_series.html

Any ideas?
bytelinker

**Comment by newville** [comment]
@bytelinker Well, assuming that I understand correctly, we should point out that a Fourier series is a linear problem, not a non-linear one:
```  
     f(x) = c_0 + Sum_n=1^N ( c_n * sin(n*x))
```
so for exactly and only "fit a Fourier series", you do not need an iterative method  -  a regression method would work, as mentioned in that StackOverflow discussion.  I believe (but am not 100% sure) that the Symfit link you give is also describing a linear least-squares fit. 

That said, it might be useful to have a lmfit model for a Fourier series.  Writing such a model for a fixed maximum number of coefficients would not be too hard.  Say, something like

```
import numpy as np
def fourier_model(x, c0, c1, c2, c3, c4, c5, c6, c7, c8, c9):
      return (c0 + c1*np.sin(x) + c2*np.sin(2*x) + c3*np.sin(3*x) + 
                   c4*np.sin(4*x) + c5*np.sin(5*x) + c6*np.sin(6*x) + 
                   c7*np.sin(7*x) + c8*np.sin(8*x) + c9*np.sin(9*x) )

```

That would allow up to 9 variable parameters, and you could set any of them to 0 to reduce the number of components.
Extending that to 25 is probably not unreasonable, though extending it to 1000 is probably not.  
An iterative approach is also just unlikely to work with a large number of components unless you have very good guesses, so (again) a linear solution might be an important first step. 

For a truly "variable" number of components, I think we would have to write a slightly more elaborate sub-class of `lmfit.Model`, as we did for `SplineModel`.   If we did that (suggestions and contributions welcome), I would suggest that we have a  `guess` method that does the linear regression for the initial values of the coefficients.
